rs/messaging/src/routing/stream_handler.rs

```diff
@@ -766,9 +763,28 @@ impl StreamHandlerImpl {
                     *available_guaranteed_response_memory -=
                         stream.push(generate_reject_response(&request, code, err.to_string()))
                             as i64;
-                    stream.push_accept_signal()
+                    stream.push_accept_signal();
+                }
+                Some((err, RequestOrResponse::Request(_))) => {
+                    // Unable to induct a request, push a reject signal.
+                    let reason = match err {
+                        StateError::CanisterMigrating { .. } => RejectReason::CanisterMigrating,
+                        StateError::CanisterNotFound(_) => RejectReason::CanisterNotFound,
+                        StateError::CanisterStopped(_) => RejectReason::CanisterStopped,
+                        StateError::CanisterStopping(_) => RejectReason::CanisterStopping,
+                        StateError::QueueFull { .. } => RejectReason::QueueFull,
+                        StateError::OutOfMemory { .. } => RejectReason::OutOfMemory,
+                        // Unreachable.
```

Comment:
I know I'm late to the party, and maybe I'm repeating a discussion from two months ago (sorry if I'm doing that), but: why not have `induct_impl` return a `RejectReason` and then do the conversion, getting rid of the unreachable stuff?

Reply:
That's how I had it before I realized we have to support certification version 18 and below too, i.e. generate reject responses rather than signals. This is implemented by turning `StateError` into a string which is why I went with  returning an the error, not a signal.

OTOH you could perhaps change that to generate a reject response from the signal rather than the error but the error contains more information than a signal so I can't produce the exact same message. I'll give that a try; if it doesn't significantly bloat the diff or causes problems in execution or something this might be the cleaner solution.

Reply:
So when I initially put support for older certification versions back in, I made it exactly the way it was before. But that is actually not necessary; I'm not sure why I did it like that TBH.

I changed it now such that `induct_message_impl()` returns `Option<(RejectReason, RequestOrResponse)>` and thereby moved the conversion from `StateError` to `RejectReason` in there directly (as it was at an earlier stage). So what happens
- for certification version 19 and after is that it just uses this `RejectReason` to push a `RejectSignal`. On the other subnet it then eventually generates and inducts a reject response locally when gc'ing the corresponding request using the reject signal to identify the proper pair of `reason` and ` request` using `generate_reject_response_for(reason, request)`.
- for certification version 18 and below it uses the `reason` and `request` in place in `induct_message()` on this subnet to generate a reject response using `generate_reject_response_for(reason, request)` and then pushes that into the stream to the other subnet.

So from the point of view of a canister on the other subnet I get exactly the same reject response for all certification versions. However note that this PR slightly changes the message contained in the reject response for all certifications versions (before I changed it only for version 19 at up). But this was discussed and is not a problem as far as I know.

For example for an 'out of memory' reject response, we previously included how much memory was left on the subnet because this information is available in `StateError::OutOfMemory`. But since a reject signal is just a stream index and a reject reason, this information is lost and thus not included anymore (note: it still appears in the logs). However it was quite useless anyway because a request always consumes 2MB due to the reservation so it's not like a canister could have constructed a smaller request to fit it in (and also the situation would have changed in the mean time anyway).

This change allowed me to remove `StateError::CanisterMigrating` (which I added some time ago) since the only place it was used is to return it from `induct_message_impl()` and it otherwise caused only some minor changes in old tests to adapt for the new way a reject response is generated. I also slightly changed the message used for logging since before it claimed to produce a reject response which is actually false depending on the certification version. It is now neutral such that it both fits generating a reject response and pushing a reject signal.

I think this is a good change and I'm fairly certain @alin-at-dfinity would agree.

Reply:
One additional note about the return type here. It was a bit weird to return `Result<(), (StateError, RequestOrResponse)>` because returning `Ok(())` can mean _successfully inducted_ or _silently_dropped_ so essentially it means _no_further_action_required_. Then I changed it to `Option<(StateError, RequestOrResponse)>` since that seems more idiomatic for this case, but then it looks a bit weird that we return an error as optional.

So long story short `Option<(RejectReason, RequestOrResponse)>` removes all the weirdness since it's not an error anymore, but it still includes the implication that the caller is expected to do something with it unless it is `None`.